### PR TITLE
[opt] load promotion in mandatory combine

### PR DESCRIPTION
**Pull request by zoecarver**
This patch adds support for load promotion in mandatory combine. This is from #28536 and will be needed (or similar) to support all the optimizations done by closure cleanup in mandatory inlining. 

For example:
```
store x to as
y = load as
use y
```
Turns into (minus copies):
```
store x to as
use x
```

I plan to also add copy/destroy and store elimination in future patches which should help reduce the dead noise left behind. 

This is one of the biggest single patches from #28536 the rest will be smaller (in size and scope). A lot of the logic here could also apply to mem2reg and #30308.

### CODE REVIEW

File: include/swift/SILOptimizer/Utils/InstOptUtils.h [hunk 1]
```diff
@@ -547,6 +547,8 @@ bool tryOptimizeApplyOfPartialApply(
     PartialApplyInst *pai, SILBuilderContext &builderCtxt,
     InstModCallbacks callbacks = InstModCallbacks());
 
+bool dominatesAllUses(SILInstruction *a, SILValue b);
```
Comment (zoecarver) [inline include/swift/SILOptimizer/Utils/InstOptUtils.h]: Implemented in #30464.

**Comment by zoecarver** [comment]
@swift-ci please benchmark.

**Comment by gottesmm** [comment]
I don't think it makes sense to do this at -Onone. It would make more sense to do a small bit of analysis to recognize the specific case. We need to be really careful at -Onone with compile time. @atrick your thoughts?

**Comment by zoecarver** [comment]
That's a good point. Two alternatives: a) only scan single blocks (which would greatly reduce the complexity here) and b) only do this for `parital_apply`s at the same time as the other optimizations. If we want to support everything from closure cleanup we would have to do (at least) the second one.

**Comment by zoecarver** [comment]
@swift-ci please smoke test compiler performance.

**Comment by zoecarver** [comment]
@swift-ci please smoke test compiler performance.

**Comment by zoecarver** [comment]
@swift-ci please smoke test compiler performance

**Comment by zoecarver** [comment]
Some todos:
- [ ] use dominance analysis instead of dominance info.
- [ ] move into another pass, not mandatory combine.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**